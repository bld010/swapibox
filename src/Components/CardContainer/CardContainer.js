import React, { Component } from 'react';
import PropTypes from 'prop-types';

class CardContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      filterSelected: 'people',
      isLoading: true
    }
  }

  componentDidMount()  {
    
  }

  render() {
    return(
      <p>This is the card container</p>
    )
  }
}


export default CardContainer;

CardContainer.propTypes = {
}