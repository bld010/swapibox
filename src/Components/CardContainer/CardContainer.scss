$background-color: #000000;

.CardContainer {
  background: $background-color;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  padding: 50px;
  min-height: 500px;
  margin: 0;
  font-family: 'Courier';
}

.CardContainer.favorites {
  font-size: 30px;
  color: green;
}