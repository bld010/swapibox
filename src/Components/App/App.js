import React, { Component } from 'react';
import './App.scss';
import Header from '../Header/Header';
import CardContainer from '../CardContainer/CardContainer';
import { fetchMovie, fetchCards } from '../Fetch/Fetch';
import { Route } from 'react-router-dom';


class App extends Component {
  constructor() {
    super(); 
    this.state = {
      planets: null,
      people: null,
      movie: null,
      vehicles: null,
      favorites: [],
      isLoading: true,
      peopleFetchError: null,
      vehiclesFetchError: null,
      planetsFetchError: null,
      movieFetchError: null,
    }
    this.fetchMovie = fetchMovie;
    this.fetchCards = fetchCards;
  }

  handleFetchError = (errorType, errorMessage) => {

    this.updateAppState(errorType, errorMessage)
  }

  updateAppState = (stateProperty, stateValue) => {
    this.setState({
      [stateProperty]: stateValue
    })
  }

  handleFavorite = (favoritedCard) => {
    if (!this.state.favorites.find(favorite => favorite.name === favoritedCard.name)) {
      this.addFavorite(favoritedCard)
    } else {
      this.removeFavorite(favoritedCard)
    }
  }

  addFavorite = (newFavoriteCard) => {
      this.setState({
        favorites: [...this.state.favorites, newFavoriteCard]
      })
  } 

  removeFavorite = (cardToRemove) => {
    this.setState({
      favorites: this.state.favorites.filter(favorite => favorite.name !== cardToRemove.name)
    })
  }

  componentDidMount() {
    this.fetchMovie(this.updateAppState, this.handleFetchError);
    this.fetchCards('https://swapi.co/api/people/', this.updateAppState, 'people', this.handleFetchError)
    this.fetchCards('https://swapi.co/api/planets/',  this.updateAppState, 'planets', this.handleFetchError)
    this.fetchCards('https://swapi.co/api/vehicles/', this.updateAppState, 'vehicles', this.handleFetchError)
  }
  

  render() {
    return (
        <div className='App'>
          <Header 
            favoritesCount={this.state.favorites.length} 
            movie={this.state.movie}  
            error={this.state.movieFetchError}
            />

          <Route exact path="/" render={() => {
            return (<section className='CardContainer'>
              <h2>Please select a link above to display people, planets, vehicles, or favorites.</h2>
            
            </section>)
          }} />

          <Route path='/people' render={() => {
            return <CardContainer 
              favorites={this.state.favorites} 
              handleFavorite={this.handleFavorite} 
              cards={this.state.people} 
              error={this.state.peopleFetchError}
              type='people' />
          }} />

          <Route path='/planets' render={() => {
            return <CardContainer 
              favorites={this.state.favorites} 
              handleFavorite={this.handleFavorite} 
              cards={this.state.planets} 
              error={this.state.planetsFetchError}
              type='planets' />
          }} />

          <Route path='/vehicles' render={() => {
            return <CardContainer 
            favorites={this.state.favorites} 
            handleFavorite={this.handleFavorite} 
            cards={this.state.vehicles} 
            error={this.state.vehiclesFetchError}
            type='vehicles' />
          }} />

          <Route path='/favorites' render={() => {
            if (this.state.favorites.length === 0) {
              return <section className='CardContainer favorites'>
                  <p>No favorites found.  Click on the circle next to the name of the person, 
                  vehicle, or planet you'd like to favorite.</p>
                </section>
            } else {
                return <CardContainer 
                  favorites={this.state.favorites} 
                  handleFavorite={this.handleFavorite} 
                  cards={this.state.favorites} 
                  type='favorites' />
            }
          }} />

        </div>
    )
    }

}

export default App;
