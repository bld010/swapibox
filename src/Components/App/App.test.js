import React from 'react';
import App from './App';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import CardContainer from '../CardContainer/CardContainer';

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });


describe('App', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<App />);

    expect(wrapper).toMatchSnapshot();
  })

  it('has a method to updateState -- passed to child components', () => {
    const wrapper = shallow(<App />);
    const initialState = wrapper.state();
    const mockPeopleArray = [{name: 'Anakin'}, {name: 'Chewy'}]

    wrapper.instance().updateAppState('people', mockPeopleArray)

    const expectedState = wrapper.state();

    expect(expectedState.people).not.toEqual(initialState.people)
    expect(expectedState.people).toEqual(mockPeopleArray)
  }
  )
  
  it('should update state with fetch errors', () => {
    const wrapper = shallow(<App />)

    const initialError = wrapper.state().peopleFetchError;
    wrapper.instance().handleFetchError('peopleFetchError', 'Error with people fetch')
    const expected = wrapper.state().peopleFetchError;
    
    expect(initialError).toEqual(null);
    expect(expected).toEqual('Error with people fetch')
  })

  it('should add favoriteCards to state', () => {
    const wrapper = shallow(<App />);
    wrapper.setState({favorites: [{name: 'Anakin'}]})
    const expected = [{name: 'Anakin'}, {name: 'Chewy'}];
    wrapper.instance().addFavorite({name: 'Chewy'})
    expect(wrapper.state().favorites).toEqual(expected);
  })

  it('should remove unfavoritedCards from state', () => {

    const wrapper = shallow(<App />);
    wrapper.setState({favorites: [{name: 'Anakin'}, {name: 'Chewy'}]});
    wrapper.instance().removeFavorite({name: 'Chewy'});
    const expected = [{name: 'Anakin'}];
    expect(wrapper.state().favorites).toEqual(expected);

  })

  describe('handleFavorite' , () => {
    it('should fire removeFavorite when new favorite is passed as argument', () => {
      const wrapper = shallow(<App />);
      const mockAddFavorite = jest.fn();
      wrapper.instance().addFavorite = mockAddFavorite;
      wrapper.setState({favorites: [{name: 'Anakin'}, {name: 'Chewy'}]});
      wrapper.instance().handleFavorite({name: 'Han'});
      expect(mockAddFavorite).toHaveBeenCalled();
    })

    it('should fire removeFavorite when current favorite is passed as argument', () => {
      const wrapper = shallow(<App />);
      const mockRemoveFavorite = jest.fn();
      wrapper.instance().removeFavorite = mockRemoveFavorite;
      wrapper.setState({favorites: [{name: 'Anakin'}, {name: 'Chewy'}]});
      wrapper.instance().handleFavorite({name: 'Anakin'});
      expect(mockRemoveFavorite).toHaveBeenCalled();
    })

  })


  describe('componentDidMount', () => {
    it('should fire 4 fetch calls', async () => {
      jest.spyOn(window, "fetch")
      const wrapper = await shallow(<App />)
      await wrapper.update()
      expect(window.fetch).toHaveBeenCalledTimes(4);
    }) 
})

describe('Router', () => {
  it('should show People component for /people router', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/people']}>
        <App />
      </MemoryRouter>
    )

    const cardContainer = wrapper.find(CardContainer)
    
    expect(cardContainer).toHaveLength(1)
  });

  it('should show Planets component for /planets router', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/planets']}>
        <App />
      </MemoryRouter>
    )

    const cardContainer = wrapper.find(CardContainer)
    
    expect(cardContainer).toHaveLength(1)
  });
  it('should show Vehicles component for /vehicles router', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/vehicles']}>
        <App />
      </MemoryRouter>
    )

    const cardContainer = wrapper.find(CardContainer)
    
    expect(cardContainer).toHaveLength(1)
  });

  it('should show prompt  for / router', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/']}>
        <App />
      </MemoryRouter>
    )

    const expected = wrapper.find('h2').props().children
    
    expect(expected).toEqual('Please select a link above to display people, planets, vehicles, or favorites.')

  });

})

})

