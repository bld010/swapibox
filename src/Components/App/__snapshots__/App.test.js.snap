// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <Header
    error={null}
    favoritesCount={0}
    movie={null}
  />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    path="/people"
    render={[Function]}
  />
  <Route
    path="/planets"
    render={[Function]}
  />
  <Route
    path="/vehicles"
    render={[Function]}
  />
  <Route
    path="/favorites"
    render={[Function]}
  />
  <Route
    render={[Function]}
  />
</div>
`;
